### PR TITLE
type config with `None` default as `str | None` instead of `Any`

### PR DESCRIPTION
**Pull request by chdsbd**
A config with a default of `None` now returns `str | None` instead of `Any`.

```python
from starlette.config import Config
config = Config()
PRIVATE_KEY = config("PRIVATE_KEY", default=None)

# before
reveal_type(PRIVATE_KEY) # Revealed type is "Any"

# after
reveal_type(PRIVATE_KEY) # Revealed type is "Union[builtins.str, None]"
```

Initially raised as discussion https://github.com/encode/starlette/discussions/1730

### CODE REVIEW

File: tests/test_config.py [hunk 1]
```diff
@@ -1,12 +1,80 @@
...
 from pathlib import Path
+from typing import Any
 
 import pytest
 
 from starlette.config import Config, Environ, EnvironError
 from starlette.datastructures import URL, Secret
 
 
+def test_config_types() -> None:
```
Comment (chdsbd) [inline tests/test_config.py:12]: I added some tests for the type annotations. Open to feedback on how to make these nicer

File: tests/test_config.py [hunk 2]
```diff
@@ -1,12 +1,80 @@
...
+    is_str(STR_CAST_STR)
+
+    BOOL = [config("BOOL", cast=bool)]
+    is_bool(BOOL)
+    BOOL_DEFAULT = [config("BOOL_DEFAULT", cast=bool, default=False)]
+    is_bool(BOOL_DEFAULT)
+    BOOL_NONE = [config("BOOL_NONE", cast=bool, default=None)]
+    is_bool_or_none(BOOL_NONE)
+
+    # defaults with non-bool types are allowed by our type annotations.
```
Comment (chdsbd) [inline tests/test_config.py]: See https://github.com/encode/starlette/discussions/1731

In a future PR, I'd like to require that the default be either `None` or the same type as the `cast`.

File: tests/test_config.py (hunk 1 above)
Comment (Kludex) [reply tests/test_config.py:12]: Maybe parametrize those:
```python
STR = [config("STR")]
is_str(STR)
```
And move the function definitions outside this test function (above) 🤔

**Review by Kludex — COMMENTED** [review]
I'm not sure if we should take an approach in which we do assertions instead of depending on the type checker.

I've asked around how people do this, not sure if helpful, but there are some ideas: https://twitter.com/marcelotryle/status/1544374655481532422

File: setup.py [hunk 3]
```diff
@@ -38,7 +38,7 @@ def get_long_description():
     include_package_data=True,
     install_requires=[
         "anyio>=3.4.0,<5",
-        "typing_extensions>=3.10.0; python_version < '3.10'",
+        "typing_extensions>=4.2.0; python_version < '3.10'",
```
Comment (Kludex) [inline setup.py]: Please undo this 😗 
```suggestion
        "typing_extensions>=3.10.0; python_version < '3.10'",
```

File: tests/test_config.py [hunk 4]
```diff
@@ -1,12 +1,43 @@
...
+from typing import Any, Optional
 
 import pytest
+from typing_extensions import assert_type
 
 from starlette.config import Config, Environ, EnvironError
 from starlette.datastructures import URL, Secret
 
 
+def test_config_types() -> None:
```
Comment (Kludex) [inline tests/test_config.py:12]: Maybe adding a docstring explaining a bit that we are relying on mypy? 🙏

**Review by Kludex — COMMENTED** [review]
I added some comments. I'm overall fine with this PR. 🙏

File: requirements.txt [hunk 5]
```diff
@@ -9,6 +9,7 @@ databases[sqlite]==0.5.5
 flake8==3.9.2
 isort==5.10.1
 mypy==0.961
+typing_extensions==4.2.0
```
Comment (chdsbd) [inline requirements.txt:12]: I added typing_extensions here because we need a newer version that contains assert_type

Comment (Kludex) [reply requirements.txt:12]: Thanks 🙏

**Review by Kludex — APPROVED** [review]
Thanks @chdsbd 🙌

I'll let merge this in a few days if there are no further comments from other members. 😁

File: tests/test_config.py [hunk 6]
```diff
@@ -1,12 +1,46 @@
+from __future__ import annotations
+
```
Comment (Kludex) [inline tests/test_config.py]:
```suggestion
```

**Pull request merged**